README.md
WebServer
=========

An efficient server library for quickly creating a WebServer and handling HTTP requests, WebSocket
connections, and routing API requests using the Dart language.

Includes extra nice features, such as setting a parameter to require Basic Authentication for a Url,
with all of the difficult auth checking and responding taken care of by the server, plus much more!

#### Who is using this package?

__[SocialFlare](https://socialflare.us/)__
* Used for preprocessing and serving webpages and resources, serving API responses, and multiple other features.
* __Major backer:__ Guarantees long-term support for this project's concept. Thank you, SocialFlare!

__[Ebates, Inc.](http://www.ebates.com/)__
* For a handful of internal tools for organizing data and serving stat pages.
* Used to serve a realtime Purchases Stat information webpage to merchant representatives at the __Ebates MAX Conference__
  built around this package.

Use Example (for no coding needed)
----------------------------------

You can use this WebServer to serve files without even having to code a single line. Using Dart's
`pub global activate` feature, you can add the WebServer package as an executable to call from the
command line in any directory and serve files from it.

~~~bash
# Activate the WebServer globally.
pub global activate web_server
~~~

~~~bash
# Navigate to the Directory that you want to serve files from.
cd /path/to/directory

# Activate the WebServer on that Directory. Defaults to port 8080.
# May require 'sudo' on Mac/Linux systems to bind to port 80.
# For: http://127.0.0.1:9090/path/to/file
web_server --port=9090
~~~

For details on all of the possible arguments and uses:
~~~bash
# Use the 'help' argument
web_server --help
~~~

__Please don't forget to run the__ `pub global activate web_server` __command every once and a while__
to get the latest version of the WebServer package; Pub/Dart does not automatically update the package to
avoid the risk of breaking changes.

Feel free to view the [CHANGELOG](CHANGELOG.md) before updating for documentation about whenever there is
a __breaking change__. Skim quickly by looking for the bold text "__Breaking Change:__" before a
"Tools > web_server" category change. It is safe to assume there will NOT be a breaking change unless the
version number increases by 2.x; the 2.0+x format changes, for example, are non-breaking when the number
after "+" is the only difference.

Features & Use Example (for coders)
-----------------------------------

Please check out the ["example/"](example/) folder in this package for full details.

### For preprocessing HTML like PHP

Use Angular-like variables, which will be converted using a helper method from this package (see Dart
code below).

__web/index.html__
~~~html
<body>
  <h1>Welcome, {{username}}!</h1>

  <p>The date today is: {{todayDate}}.</p>
</body>
~~~

Then, process variables like PHP on the Dart server side:

__server.dart__

~~~dart
import "dart:io";
import "package:html/parser.dart" as domParser; // https://pub.dartlang.org/packages/html
import "package:html/dom.dart" as dom;
import "package:web_server/web_server.dart" as webServer;

void main() {
  // Initialize the WebServer  
  final webServer.WebServer localWebServer = new webServer.WebServer(InternetAddress.ANY_IP_V4, 8080,
        hasHttpServer: true);
        
  localWebServer.httpServerHandler
    .forRequestPath(new webServer.UrlPath('/index.html')).onRequested.listen((final HttpRequest httpRequest) async {
      String indexFileContents = await new File('path/to/index.html').readAsString();
      
      // Apply the Dart variables to the HTML file's variables like
      // AngularJS/AngularDart
      indexFileContents = webServer.applyVarModifiers(indexFileContents, {
        "username": "mrDude",
        "todayDate": '${new DateTime.now()}'
      });
      
      // ===== AND/OR =====
      // Interact with the HTML like client side Dart.
      final dom.Document document = domParser.parse(indexFileContents);
      
      // The HTML library has its own Element Objects; separate from the 'dart:html' ones.
      final dom.Element pElm = document.querySelector('p');
      
      pElm.remove(); // Remove the <p> Element from the document's DOM.
      
      // Add data to and close out the Http request's response.
      httpRequest.response
          ..headers.contentType = new ContentType('text', 'html', charset: 'utf-8')
          
          ..write(indexFileContents)
          // OR
          ..write(document.outerHtml)
          
          ..close();
    });
}
~~~

### For Hosting APIs

Filter every request starting with a certain Url pattern into a request handler.

~~~dart
import "dart:io";
import "package:web_server/web_server.dart" as webServer;

void main() {
  // Initialize the WebServer
  final webServer.WebServer localWebServer = new webServer.WebServer(InternetAddress.ANY_IP_V4, 8080,
        hasHttpServer: true);
        
  localWebServer.httpServerHandler
      // NOTE: ApiHandler would be a Class or namespace created by you in your code, for example.
      ..handleRequestsStartingWith(new webServer.UrlPath('/api/categories')).listen(ApiHandler.forCategories)
      ..handleRequestsStartingWith(new webServer.UrlPath('/api/products')).listen(ApiHandler.forProducts)
      ..handleRequestsStartingWith(new webServer.UrlPath('/api/users')).listen((final HttpRequest httpRequest) {
        // Create the Object for the response
        final webServer.ApiResponse apiResponse = new webServer.ApiResponse()
            ..addData("username", "mrDude") // Add data
            ..addData("email", "radical_surfer@example.com")
            ..addData("userId", 1425302);

        // Send the data back through to the request
        httpRequest.response
            // Set to "application/json; charset=utf-8"
            ..headers.contentType = ContentType.JSON
            
            // Stringify the JSON output, then send to client
            ..write(apiResponse.toJsonEncoded())
            
            ..close();
      });
}
~~~

### Static File Directory/Basic WebServer

~~~dart
import "dart:io";
import "dart:async";
import "package:web_server/web_server.dart" as webServer;

Future<Null> main() async {
  // Initialize the WebServer  
  final webServer.WebServer localWebServer = new webServer.WebServer(InternetAddress.ANY_IP_V4, 8080,
        hasHttpServer: true);
        
  // Attach HttpServer pages and event handlers
  await localWebServer.httpServerHandler
      // Automatically recursively parse and serve all items in this
      // directory matching the accepted file types (optional parameter).
      .serveStaticVirtualDirectory('web',
          supportedFileExtensions: const <String>['html', 'css', 'dart', 'js'], // Optional restriction
          shouldPreCache: true);
}
~~~

### WebSocket Server

Let the WebServer automatically handle upgrading and connecting to WebSockets from the client
side. The WebServer will forward data related to important events and automatically call your
event listeners if you send data through a WebSocket from the client with the "cmd" parameter
in the payload's Map Object.

~~~dart
import "dart:io";
import "package:web_server/web_server.dart" as webServer;

void main() {
  // Initialize the WebServer with the hasWebSocketServer parameter
  final webServer.WebServer localWebServer = new webServer.WebServer(InternetAddress.ANY_IP_V4, 8080,
          hasHttpServer: true, hasWebSocketServer: true);
          
  // HTTP Server handlers code here...
  
  // Attach WebSocket command listeners and base events
  localWebServer.webSocketServerHandler
      // For automatically routing handling of data sent through a WebSocket with this pattern of "cmd":
      // {"cmd": 0, "data": { "pokemonCount": 151 }}
      ..on[0].listen((final webServer.WebSocketRequestPayload requestPayload) { /*...*/ })
      ..onConnectionOpen.listen((final webServer.WebSocketConnectionData connectionData) { /*...*/ })
      ..onConnectionError.listen((final WebSocket webSocket) { /*...*/ })
      ..onConnectionClose.listen((final WebSocket webSocket) { /*...*/ });
}
~~~

### Add a custom ContentType

Allows the server to automatically pick up on this file extension as the supplied ContentType parameter
when it is handling serving files.

~~~dart
HttpServerRequestHandler.addContentType('.html', new ContentType('text', 'html', charset: 'utf-8'));
~~~

Features and bugs
-----------------

Please file feature requests and bugs using the GitHub issue tracker for this repository.

Using this package? Let me know!
--------------------------------

I am excited to see if other developers are able to make something neat with this package.
If you have a project using it, please send me a quick email at the email address listed on
[my GitHub's main page](https://github.com/bwhite000). Thanks a bunch!


CHANGELOG.md
WebServer Changelog
===================

v2.0.0+3 (3.5.2016)
-------------------

### Tool Changes

* `web_server` - pub global
  * The developer can now pass a `--port=<port-number>` argument to the executable in the command line to
    specify a specific port, not just the default 8080, for the web server to bind itself to; e.g.
    `web_server --port=80`
  * Added an interpretation of the `--help` or `-h` argument to output details into the terminal about the
    executable and its functionality.
  * Minor understandability improvement to the output terminal log when the server is started.
  * Added another optional developer command line argument for `--host=<address>` to allow binding to a
    specific address, such as only `127.0.0.1`; the default value is `0.0.0.0`.
    
### Documentation Changes
* ReadMe
  * Changed the filename to uppercase to be more inline with README file naming formats.
  * Updated the examples to reflect the new parameters in the web_server tool.
  * Added the very important details about reminding developers to run `pub global activate` on the package
    every once and a while to get the latest updates; also included details about easily checking for breaking
    changes before updating; hopefully the details will make it not-so-scary to update.
* ChangeLog
  * Changed the filename to uppercase to be more inline with CHANGELOG file naming formats.

v2.0.0+2 (1.13.2016)
--------------------

### Tool Changes

* `web_server` - pub global
  * `shouldPreCache` has been changed to false since the Dart language seems to not be closing out filesystem
    connections, even when explicitly coded to, and will error when setting up to serve directories with more
    files than the computer's maximum file connection limit; will now cache as the files are requested instead
    of in advance.

v2.0.0+1 (12.22.2015)
---------------------

### Library Changes
  
* `HttpServerRequestHandler`
  * Added better comments to some methods for the DartDocs generator to make better use of.
  * Removed the need for the '.' in the supportedFileExtensions extension list in `serveStaticVirtualDirectory`
    that was unintentionally introduced in the previous release; functionality still behaved like before,
    but the dot was needed; not anymore now.

### Documentation/Example Changes

* ReadMe
  * Updated some code examples and section information to be easier to understand.
  * Added a code example of preprocessing using the `html` Dart package to show that a developer can modify
    web pages as richly and complexly as they could on the client side, but before even serving the page; this
    way, the developer can have code templates in their web page's HTML, use it on the server like client side
    DOM to fill it with data and clone it, if wanted, then return the built page to the client.
  
* Code Example Files
  * Cleaned-up and fixed the code to be efficiently and aesthetically better; e.g. added the `final` keyword
    in a few missing spots.

* Changelog
  * Made the markdown data easier to read by dividing common change areas into sections; inspired by the
    Changelog pattern in the main Dart SDK GitHub.

v2.0.0 (12.20.2015)
-------------------

### Library Changes

* `WebServer`
  * Added a new constructor for binding a secure server and switched the syntax to using the new Dart 1.13
  BoringSSL format instead of the old NSS method; use the `new WebServer.secure()` constructor to support this.
  * Removed the 'GET' and 'POST' allowedMethods default value so that any type of request header will be
    allowed by default; the developer will no longer have to specify every type of allowed request method.
  * Bumped the default wait time for a response to generate from the server code (response deadline) from
    20 seconds to 30 seconds to allow for more complex response generation to not catch the developer off
    guard as quickly in case the developer isn't coding with this deadline in mind.

* `HttpServerRequestHandler`
  * Added content types for .mp3, .ogg, .oga, .ogv.
  * Changed the \_fileExtensions content types over to using the ContentType Object instead of a List<String>
    and having to build the ContentType Object for every request; also, removed the declaration of it being a
    'const' so that a developer using the server would be able to add their own new content types
    programatically.
  * Some code clean-up and optimizations for the occasional variable reuse optimization or the similar.
  * Created a new format for listening for webpage path requests to make it easier to understand and more
    intuitive by making the event format similar to binding DOM events on a webpage, like, for example, the
    format of `querySelector('...').onClick.listen((_) {})`.
  * Added a method and structures for opening the ability for developers using this package to add their own
    custom error code handlers using `.onErrorDocument()`.
  * Deprecated `serveVirtualDirectory()` in favor of the clearer and having more features, like being dynamic,
    `serveStaticVirtualDirectory()` and `serveDynamicVirtualDirectory()` (arriving eventually; a.k.a. coming
    soon).
  * Renamed the `UrlData` Class to `UrlPath` to make it easier to understand what the Object is representing.
  * `serveStaticVirtualDirectory`
    * Made the requirement of providing a whitelist of supported file extensions
      an optional parameter to allow for serving an entire directory, or a directory with flexible file types,
      simple and not requiring a server restart; also, this makes it possible for the pub global `web_server`
      command to operate with any directory's files.
    * Removed a loop that was checking file extensions on every file entity for a match in the
      `serveStaticVirtualDirectory` and is now doing a `.contains()` on the List to be much faster and loop
      less.
    * Removed the redundant `isRelativePath` parameter.

example/web_server_misc.dart
import "dart:io";
import "package:web_server/web_server.dart";

void main() {
  // Initialize and bind the HTTP and WebSocket WebServer
  final WebServer localWebServer = new WebServer(InternetAddress.ANY_IP_V4, 8080,
      hasHttpServer: true, hasWebSocketServer: true);

  // Log out some of the connection information
  stdout.writeln('Local web server started at: (http://${localWebServer.address.address}:${localWebServer.port})'); // http://127.0.0.1:8080

  HttpServerRequestHandler.shouldBeVerbose = true;

  // Attach HttpServer pages and event handlers
  localWebServer.httpServerHandler
      // Gain handling of navigations to "/index.html"
      ..forRequestPath(new UrlPath('/index.html')).onRequest.listen((final HttpRequest httpRequest) { /*...*/ })

      // Gain handling to ANY immediate sub-item in the directory;
      // serveVirtualDirectory() is preferred over this unless you need fine grain controls
      ..registerDirectory(new UrlPath('/img/profile_pics/80/')).listen((final HttpRequest httpRequest) { /*...*/ })

      // Automatically parse for indexing and serve all recursive items in this directory matching the accepted file extensions.
      ..serveStaticVirtualDirectory('lib/web_demo', supportedFileExtensions: const <String>['html', 'css', 'dart', 'js'])

      // Automatically handle serving this file at navigation to '/static_page', with optional in-memory caching
      ..serveStaticFile(new UrlPath('/static_page'), 'web_demo/static_page.html', enableCaching: false)

      // Gain handling of all API requests, for example; catches all paths starting with the String in UrlData
      ..handleRequestsStartingWith(new UrlPath('/api/')).listen((final HttpRequest httpRequest) {/*...*/})

      // Handle requiring Basic Authentication on the specified Url, allowing only the users in the authentication list.
      // The required credentials are "user:password" (from the BasicAuth base64 encoded -> 'dXNlcjpwYXNzd29yZA==')
      ..registerPathWithBasicAuth(new UrlPath('/api/auth/required/dateTime'), const <AuthUserData>[
            const AuthUserData('username', 'dXNlcjpwYXNzd29yZA==') // user:password --> Base64
          ]).listen((final HttpRequest httpRequest) {
            // Create a new ApiResponse object for returning the API data;
            // Value --> {"sucess": true, "dateTime": "XXXX-XX-XX XX:XX:XX.XXX"}
            final ApiResponse apiResponse = new ApiResponse()
                ..addData('dateTime', '${new DateTime.now()}'); // Add the DateTime

            // Respond to the Url request
            httpRequest.response
                ..headers.contentType = ContentType.JSON // Set the 'content-type' header as JSON
                ..write(apiResponse.toJsonEncoded()) // Export as a JSON encoded string
                ..close();
          })

      // Add a custom function for handling the request in case of the error code supplied as the parameter.
      ..onErrorDocument(HttpStatus.NOT_FOUND, (final HttpRequest httpRequest) {
        httpRequest.response
            ..statusCode = HttpStatus.NOT_FOUND
            ..headers.contentType = new ContentType('text', 'html', charset: 'utf-8')
            ..write('<h1>404 Error accessing: ${httpRequest.requestedUri.path}</h1>')
            ..close();
      });

  // Attach WebSocket command listeners and base events
  localWebServer.webSocketServerHandler
      ..on[0].listen((final WebSocketRequestPayload requestPayload) { /*...*/ })
      ..onConnectionOpen.listen((final WebSocketConnectionData connectionData) { /*...*/ })
      ..onConnectionError.listen((final WebSocket webSocket) { /*...*/ })
      ..onConnectionClose.listen((final WebSocket webSocket) { /*...*/ });
}